### PR TITLE
Ingester main loop

### PR DESCRIPTION
**Pull request by msf**
This is the main loop logic, it connects the:
 - rpc node client that is getting blocks
 - dune api client, that is sending blocks to dune.

it has a few responsabilities:
- track errors on either side (and later do something about it)
- track if it is behind or up to date
- try to go fast.
- produce good logs of whats going on

### CODE REVIEW

File: ingester/mainloop.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	return nil
+}
+
+func (i *ingester) SendBlocks(ctx context.Context, blocksCh <-chan models.RPCBlock) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case payload, ok := <-blocksCh:
+			// TODO: we should batch RCP blocks here before sending to Dune.
```
Comment (vegarsti) [inline ingester/mainloop.go:120]: Something like this?

```
	// Every N seconds: Send the batched blocks
	buffer := make([]*Block, 0, concurrency)
	ticker := time.NewTicker(batchInterval)
	for {
		select {
		case block := <-blocks:
			buffer = append(buffer, block)
			if len(buffer) == cap(buffer) {
				sendBlocks(buffer)
				buffer = buffer[:0] // Clear the buffer after sending.
			}
		case <-ticker.C:
			if len(buffer) > 0 {
				sendBlocks(buffer)
				buffer = buffer[:0] // Clear the buffer after sending.
			}
		}
	}
```

File: ingester/mainloop.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	latest, err := i.node.LatestBlockNumber()
+	if err != nil {
+		i.log.Error("Failed to get latest block number, continuing..", "error", err)
+		return atomic.LoadInt64(&i.info.LatestBlockNumber)
+	}
+	atomic.StoreInt64(&i.info.LatestBlockNumber, latest)
+	return latest
+}
+
+func (i *ingester) ReportProgress(ctx context.Context) error {
```
Comment (vegarsti) [inline ingester/mainloop.go:150]: Running this in a separate goroutine is such a great idea!

**Review by vegarsti — APPROVED** [review]
Nice!!

**Comment by vegarsti** [comment]
For the main loop, I'm thinking we might want to use more of a pipeline with channels
- one channel to send block numbers on. 2 goroutines to send block numbers into this channel: one for incrementing the next block number, another for checking Dune API for missing block numbers. Or it reads from the DLQ and sends those blocks on the channel
- one channel to send _blocks_ on. Many goroutines fetch blocks and send into this channel.
- one goroutine reads blocks, buffers, and sends to the API

**Review by helanto — APPROVED** [review]
As far as I can tell, LGTM :)

File: ingester/ingester.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+	// This is just a placeholder for now
+	Info() Info
+}
+
+const defaultMaxBatchSize = 1
+
+type Config struct {
+	MaxBatchSize     int
+	StartBlockHeight int64
```
[the remark this answers is not in the document]
Comment (msf) [reply ingester/ingester.go:37]: this is to test stuff by hand or to backffil in a simple way..

File: ingester/mainloop.go (hunk 1 above)
Comment (msf) [reply ingester/mainloop.go:120]: exactly!!

File: ingester/mainloop.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+				"blockNumber", blockNumber,
+				"latestBlockNumber", latestBlockNumber,
+			)
+			time.Sleep(i.cfg.PollInterval)
+			latestBlockNumber = i.tryUpdateLatestBlockNumber()
+		}
+		return latestBlockNumber
+	}
+
+	for blockNumber := startBlockNumber; dontStop || startBlockNumber <= endBlockNumber; blockNumber++ {
```
[the remark this answers is not in the document]
Comment (msf) [reply ingester/mainloop.go:69]: agreed, like Vegard proposed above

**Pull request merged**